tools/packnam/packnam.1
.\"     Title: packnam
.\"    Author: Kent Hansen
.\" Generator: DocBook XSL Stylesheets v1.71.1 <http://docbook.sf.net/>
.\"      Date: 03/31/2008
.\"    Manual: 
.\"    Source: 
.\"
.TH "PACKNAM" "1" "03/31/2008" "" ""
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
packnam \- encodes a raw NES nametable
.SH "SYNOPSIS"
.HP 8
\fBpacknam\fR [\fBoption\fR...] {file}
.SH "DESCRIPTION"
.PP

\fBpacknam\fR
reads a raw NES nametable (merely an array of tile indexes \-\- bytes) and encodes it in a simple, space\-efficient format that can be readily decoded by 6502 code (i.e. so the graphic that the nametable represents can be displayed on the NES).
.PP
As a basic illustration of what packnam does, assume that the input is a nametable containing the following tile indexes (byte values):
.PP
0 0 0 1 1 2 3 4 1 1 1 1 1 3 3 0 0 0 0 1 0 1 0 4
.PP
The nametable is 24 bytes in size. Assume it represents a graphic consisting of 6x4 tiles (6*4=24); packnam should thus be invoked with the option
\-\-width=6. packnam will produce an encoded nametable that represents the graphic as a series of rows:
.PP
row 0: 0 0 0 1 1 2
.PP
row 1: 3 4 1 1 1 1
.PP
row 2: 1 3 3 0 0 0
.PP
row 3: 0 1 0 1 0 4
.SH "OPTIONS"
.PP
\fB\-\-width\fR=\fInum\fR
.RS 4
Use
\fInum\fR
as nametable width (tiles per row). If this option is not specified, the width is 32 (i.e. the full width of a NES hardware nametable).
.RE
.PP
\fB\-\-vram\-address\fR=\fInum\fR
.RS 4
Use
\fInum\fR
as VRAM start address (i.e. where the first row will be decoded to). This is used to control where in the NES hardware nametable the tile indexes should be written to; for example, to center the graphic or right\-align it. If this option is not specified, the VRAM start address will be 0x2000 (beginning of NES hardware nametable #0).
.RE
.PP
\fB\-\-output\fR=\fIfile\fR
.RS 4
Store encoded nametable in
\fIfile\fR. If this option is not specified, the data will be written to
packnam.dat.
.RE
.PP
\fB\-\-zero\-terminate\fR
.RS 4
Zero\-terminate the output (i.e. add a trailing 0\-byte).
.RE
.PP
\fB\-\-verbose\fR
.RS 4
Print progress information to standard output.
.RE
.PP
\fB\-\-help\fR
.RS 4
Give a help list.
.RE
.PP
\fB\-\-usage\fR
.RS 4
Give a short usage message.
.RE
.PP
\fB\-\-version\fR
.RS 4
Print program version.
.RE
.SH "FILES"
.SS "Input"
.PP
The input nametable is typically one produced by the
\fBpackchr\fR
tool, or by your favorite hex editor (extra points!). It's just a series of tile indexes, each index being one byte (references a tile in the range 0..255).
.SS "Output"
.PP
The output is a sequence of strings of the following format:
.PP
byte 0: high VRAM address
.PP
byte 1: low VRAM address
.PP
byte 2: bits 0\-5: length of data string (repeat count if RLE); bit 6: is data RLE? (1 = yes); bit 7: VRAM address increment (0 = 1, 1 = 32)
.PP
byte 3\-..: data. Only 1 byte if string is RLE
.PP
6502 code for decoding data of this format can be found elsewhere, e.g. in the NeoToxin sources (ppuwrite.asm). In your 6502 sources, you can then do something like:
.PP

.sp
.RS 4
.nf
  my_nametable_data:
  .incbin "mynametable.dat" ; mynametable.dat produced by packnam
  .db 0 ; data terminator

  ...

  lda #<my_nametable_data
  ldy #>my_nametable_data
  jsr write_ppu_data_at     ; call decoder function
.fi
.RE
.sp
.PP
packnam does not deal with CHR data, palette data or attribute data; you are responsible for setting up those things when displaying the image on the NES. (A simple way of setting up the palette and attribute table is by "tacking on" their data to packnam's output; use the string format described above.)
.SH "EXAMPLES"
.PP
To pack
mynametab.nam, using 6 tiles per row and putting the graphic halfway down the screen:
.PP

\fB packnam \-\-width=6 \-\-vram\-address=0x21B0 mynametab.nam \fR
.SH "SEE ALSO"
.PP
packchr(1)
.SH "AUTHOR"
.PP
\fBKent Hansen\fR
.sp -1n
.IP "" 4
Author.


tools/packnam/packnam.c
/*
    This file is part of packnam.

    packnam is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    packnam is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with packnam.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

static char program_version[] = "packnam 1.0";

/* Prints usage message and exits. */
static void usage()
{
    printf(
        "Usage: packnam [--width=NUM] [--vram-address=NUM]\n"
        "               [--output=FILE] [--zero-terminate]\n"
        "               [--verbose]\n"
        "               [--help] [--usage] [--version]\n"
        "                FILE\n");
    exit(0);
}

/* Prints help message and exits. */
static void help()
{
    printf("Usage: packnam [OPTION...] FILE\n\n"
           "packnam encodes a raw NES nametable.\n\n"
           "Options:\n\n"
           "  --width=NUM                     Use NUM as nametable width (tiles per row) (32)\n"
           "  --vram-address=NUM              Use NUM as VRAM start address (0x2000)\n"
           "  --output=FILE                   Store encoded data in FILE\n"
           "  --zero-terminate                Zero-terminate the output\n"
           "  --verbose                       Print statistics\n"
           "  --help                          Give this help list\n"
           "  --usage                         Give a short usage message\n"
           "  --version                       Print program version\n");
    exit(0);
}

/* Prints version and exits. */
static void version()
{
    printf("%s\n", program_version);
    exit(0);
}

/*
  Packs the given \a nametable and stores the result int \a out and
  its size in \a out_sz.
*/
void pack_nametable(const unsigned char *nametable, int nametable_sz,
                    int width, int vram_address, int zero_terminate,
                    unsigned char **out, int *out_sz)
{
    int out_pos = 0;
    int buf_sz = 0;
    int y = 0;
    int in_pos = 0;
    assert(nametable);
    assert(out);
    assert(out_sz);
    *out = 0;
    while (y * width < nametable_sz) {
	unsigned char rlech;
	int count;
	int count_pos;
	int x = 0;
	int state = 0;
	int start_rle = 0;
	int addr = vram_address + (y * 32);
	while (x < width) {
	    const unsigned char ch = nametable[in_pos];
	    switch (state) {
		case 0:
                    start_rle = (x + 4 <= width)
				&& (nametable[in_pos+1] == ch)
				&& (nametable[in_pos+2] == ch)
				&& (nametable[in_pos+3] == ch);
                    /* fallthrough */
		case 1:
                    if (ch == 0) {
                        /* don't encode 0s */
                        start_rle = 0;
                        ++in_pos;
                        ++x;
                        ++addr;
                        break;
                    }
                    if (out_pos + 2 >= buf_sz) {
                        buf_sz += 64;
                        *out = (unsigned char *)realloc(*out, buf_sz);
                    }
                    (*out)[out_pos++] = (unsigned char)(addr >> 8);
                    (*out)[out_pos++] = (unsigned char)(addr & 255);
                    if (start_rle) {
                        /* start of RLE run */
                        count = 4;
                        rlech = ch;
                        in_pos += 4;
                        x += 4;
                        state = 2;
                    } else {
                        /* start of non-RLE run */
                        count_pos = out_pos;
                        if (out_pos + 2 >= buf_sz) {
                            buf_sz += 64;
                            *out = (unsigned char *)realloc(*out, buf_sz);
                        }
                        (*out)[out_pos++] = 0; /* count backpatched later */
                        count = 1;
                        (*out)[out_pos++] = ch;
                        ++in_pos;
                        ++x;
                        state = 3;
                    }
                    break;

		case 2:
                    if (ch == rlech) {
                        ++count;
                        ++in_pos;
                        ++x;
                    }
                    if ((ch != rlech) || (count == 0x3F)) {
                        /* end RLE run */
                        if (out_pos + 2 >= buf_sz) {
                            buf_sz += 64;
                            *out = (unsigned char *)realloc(*out, buf_sz);
                        }
                        (*out)[out_pos++] = (unsigned char)(0x40 | count);
                        (*out)[out_pos++] = rlech;
                        addr += count;
                        state = 0;
                    }
                    break;

		case 3:
                    start_rle = ((x + 4 <= width)
				 && (nametable[in_pos+1] == ch)
				 && (nametable[in_pos+2] == ch)
				 && (nametable[in_pos+3] == ch));
                    if (start_rle) {
                        /* end non-RLE run */
                        (*out)[count_pos] = count;
                        addr += count;
                        state = 1;
                    } else {
                        if (out_pos + 1 >= buf_sz) {
                            buf_sz += 64;
                            *out = (unsigned char *)realloc(*out, buf_sz);
                        }
                        (*out)[out_pos++] = ch;
                        ++in_pos;
                        ++x;
                        ++count;
                        if (count == 0x3F) {
                            /* end non-RLE run */
                            (*out)[count_pos] = count;
                            addr += count;
                            state = 1;
                        }
                    }
                    break;
	    }
	}
	switch (state) {
	    case 0:
	    case 1:
                break;

	    case 2:
                /* finish RLE run */
                if (out_pos + 2 >= buf_sz) {
                    buf_sz += 64;
                    *out = (unsigned char *)realloc(*out, buf_sz);
                }
                (*out)[out_pos++] = (unsigned char)(0x40 | count);
                (*out)[out_pos++] = rlech;
                break;

	    case 3:
                /* finish non-RLE run */
                (*out)[count_pos] = count;
                break;
	}
	++y;
    }
    if (zero_terminate) {
	if (out_pos + 1 >= buf_sz) {
	    buf_sz += 1;
	    *out = (unsigned char *)realloc(*out, buf_sz);
	}
        (*out)[out_pos++] = 0;
    }
    *out_sz = out_pos;
}

/**
 * Program entrypoint.
 */
int main(int argc, char **argv)
{
    unsigned char nametable[1024];
    int nametable_sz;
    unsigned char *out;
    int out_sz;
    int width = 32;
    int vram_address = 0x2000;
    int verbose = 0;
    int zero_terminate = 0;
    const char *input_filename = 0;
    const char *output_filename = 0;
    /* Process arguments. */
    {
        char *p;
        while ((p = *(++argv))) {
            if (!strncmp("--", p, 2)) {
                const char *opt = &p[2];
                if (!strncmp("width=", opt, 6)) {
                    width = strtol(&opt[6], 0, 0);
                } else if (!strncmp("vram-address=", opt, 13)) {
                    vram_address = strtol(&opt[13], 0, 0);
                } else if (!strncmp("output=", opt, 7)) {
                    output_filename = &opt[7];
                } else if (!strcmp("zero-terminate", opt)) {
                    zero_terminate = 1;
                } else if (!strcmp("verbose", opt)) {
                    verbose = 1;
                } else if (!strcmp("help", opt)) {
                    help();
                } else if (!strcmp("usage", opt)) {
                    usage();
                } else if (!strcmp("version", opt)) {
                    version();
                } else {
                    fprintf(stderr, "packnam: unrecognized option `%s'\n"
			    "Try `packnam --help' or `packnam --usage' for more information.\n", p);
                    return(-1);
                }
            } else {
                input_filename = p;
            }
        }
    }

    if (!input_filename) {
        fprintf(stderr, "packnam: no filename given\n"
                        "Try `packnam --help' or `packnam --usage' for more information.\n");
        return(-1);
    }

    /* Read input file */
    if (verbose)
        fprintf(stdout, "reading `%s'\n", input_filename);
    {
        FILE *fp = fopen(input_filename, "rb");
        if (!fp) {
            fprintf(stderr, "packnam: failed to open `%s' for reading\n", input_filename);
            return(-1);
        }

        nametable_sz = fread(nametable, 1, 1024, fp);
        fclose(fp);
    }

    /* Compress */
    if (verbose)
        fprintf(stdout, "processing\n");
    pack_nametable(nametable, nametable_sz, width, vram_address, zero_terminate, &out, &out_sz);

    /* Write output */
    if (!output_filename)
        output_filename = "packnam.dat";
    if (verbose)
        fprintf(stdout, "writing data to `%s'\n", output_filename);
    {
        FILE *fp = fopen(output_filename, "wb");
        fwrite(out, 1, out_sz, fp);
        fclose(fp);
    }

    if (verbose)
        fprintf(stdout, "compressed size: %d%%\n", (out_sz*100) / nametable_sz);

    free(out);

    /* Success */
    return 0;
}
